abiltity to dump-to-h2 removing encryption

kidd opened the pull request:
This PR allows for all fields to be dumped to an h2 db in plaintext, using the current MB_ENCRYPTION_SECRET_KEY env var.

Note: Not using multi-statement update because h2 doesn't allow it on prepared statements: http://h2-database.66688.n3.nabble.com/Invalid-value-for-parameter-in-multi-expression-statement-td898455.html 
TODO:
- [x] add test for happy path
- [x] test some fields encrypted and some plaintext
- [x] --keep-existing + --dump-plaintext?
- [ ] naming the flag better

kidd committed:
add --dump-plaintext flag to `dump-to-h2` cmd

The new flag uses the current encryption secret key to decrypt the
encrypted values from the database while dumping and stores the
plaintext values.

kidd commented on test/metabase/cmd/dump_to_h2_test.clj:
```diff
@@ -35,3 +45,38 @@
+  [db-type db-name app-db]
+  (case db-type
+    :h2 {:subprotocol "h2"
+         :subname     (format "mem:%s;DB_CLOSE_DELAY=10" app-db)
```
app-db and db-name can be the same (any string)

camsaul commented on src/metabase/cmd/copy.clj:
```diff
@@ -279,3 +280,20 @@
+    (binding [db/*db-connection* source-jdbc-spec
+              db/*quoting-style* (mdb.conn/quoting-style source-db-type)]
+      (jdbc/with-db-connection [target-conn target-jdbc-spec]
+
```
https://guide.clojure.style/#no-blank-lines-within-def-forms

camsaul commented on test/metabase/cmd/dump_to_h2_test.clj:
```diff
@@ -35,3 +45,38 @@
+       (load-from-h2/load-from-h2! h2-fixture-db-file)
+       (eu/with-secret-key "89ulvIGoiYw6mNELuOoEZphQafnF/zYe+3vT+v70D1A="
+         (db/insert! Setting {:key "my-site-admin", :value "baz"})
+         (dump-to-h2/dump-to-h2! h2-file {:dump-plaintext? true}))
```
might be worth adding a test with `:dump-plaintext? false` here so you can see if the Setting is still encrypted in that version

camsaul replied:
you should probably also test Database `details` here as well

kidd replied:
added tests for details and set flag to false as well as test that default==false

camsaul commented on test/metabase/cmd/dump_to_h2_test.clj:
```diff
@@ -1,11 +1,21 @@
-            [metabase.cmd
-             [copy :as copy]
-             [dump-to-h2 :as dump-to-h2]]
             [metabase.cmd :as cmd]
-            [metabase.util.files :as u.files]))
+            [metabase.cmd.copy :as copy]
```
If the version of `master` you branched off of was using the nested `:require` style, it must be a little out of date -- we swtiched over 2 weeks ago I think. Maybe try merging `master` in to this PR or even rebasing the branch on `origin/master` and seeing if that gets CodeCov to stop being weird

kidd replied:
I checked and master has the nested version. maybe when the refactor was done, this file was not in the repo?

Anyway, I rebased to master just in case, but probably this diff will still appear

camsaul reviewed (approved):
Overall LGTM. Left a few minor style notes and suggestions for more test cases we should add. But you don't need me to re-review it after you address those...

Try merging in `origin/master` or rebasing off of `origin/master` and see if that fixes CodeCov. If not we can merge anyway since it looks like it's a little confused in this PR

kidd committed:
fix dump-to-h2 --dump-plaintext for database details

Database details have to be treated as plain strings instead of
relying on mapping of json to clojure maps.
